File: bishi/4-17华子/b.py

import sys

from collections import defaultdict, deque

def solve() -> None:
    m, n = map(int, input().split())
    adj = defaultdict(list)
    root = set()
    p = [defaultdict(int), defaultdict(int)]
    d = defaultdict(int)
    node = set()

    for i in range(n):
        u, v, level, cnt = input().split()
        level, cnt = int(level), int(cnt)

        if v == "*":
            root.add(u)
        else:
            adj[u].append(v)
            d[v] += 1
        p[level][u] = cnt
        node.add(u)

    ans = 0
    q = deque()
    for u in node:
            if d[u] == 0:
                q.append(u)
    
    while q:
        u = q.popleft()
        for v in adj[u]:
            p[0][v] += p[0][u]
            p[1][v] += p[1][u]
            d[v] -= 1
            if d[v] == 0:
                q.append(v)
    
    for u in root:
        if p[0][u] * 5 + p[1][u] * 2 > m:
            ans += 1
    print(ans)

if __name__ == '__main__':
    solve()

File: bishi/4-17华子/b.cpp
#include <bits/stdc++.h>

using namespace std;
int tot = 0;
int cnt[2][100000];
vector<int> adj[100000];
set<int> root, node;
int d[100000];

void solve() {
    int m, n;
    cin >> m >> n;
    map<string, int> id;


    for (int i = 0; i < n; i++) {
            string x, y;
            int l, v;
            cin >> x >> y >> l >> v;
            if (!id.count(x)) {
                    id[x] = ++tot;
            } 
            if (y == "*") {
                    root.insert(id[x]);
            }
            if (!id.count(y) && y != "*") {
                    id[y] = ++tot;
            }
            if (y != "*") {
                    d[id[y]]++;
                    adj[id[x]].push_back(id[y]);
            }
            cnt[l][id[x]] = v;
            node.insert(id[x]);
    }

    int ans = 0;

    queue<int> q;
    for (int x : node) {
            if (d[x] == 0) {
                    q.push(x);
            }
    }

    while (!q.empty()) {
            int u = q.front();
            q.pop();
            for (auto v : adj[u]) {
                    cnt[0][v] += cnt[0][u];
                    cnt[1][v] += cnt[1][u];
                    if (--d[v] == 0) {
                            q.push(v);
                    }
            }
    }

    for (auto u : root) {
        if (cnt[0][u] * 5 + cnt[1][u] * 2 > m) {
                ans++;
        }   
    }

    std::cout << ans << "\n";
    
}

int main() {
    
    solve();

    return 0;
}
